pkg/util/docker/dockerfile/builder/client.go

```diff
@@ -0,0 +1,388 @@
+	// the current working directory if not set.
+	Directory string
+	// Excludes are a list of file patterns that should be excluded
+	// from the context. Will be set to the contents of the
+	// .dockerignore file if nil.
+	Excludes []string
+	// Tag is an optional value to tag the resulting built image.
+	Tag string
+
+	// Container is optional and can be set to a container to use as
+	// the execution environment for a build.
+	Container *docker.Container
+	// Image is optional and may be set to control which image is used
+	// as a base for this build. Otherwise the FROM value from the
+	// Dockerfile is read (will be pulled if not locally present).
+	Image *docker.Image
+
+	// AuthFn will handle authenticating any docker pulls if Image
+	// is set to nil.
+	AuthFn func(name string) ([]docker.AuthConfiguration, bool)
+	// HostConfig is used to start the container (if necessary).
+	HostConfig *docker.HostConfig
+}
+
+// NewClientExecutor creates a client executor.
+func NewClientExecutor(client *docker.Client) *ClientExecutor {
+	return &ClientExecutor{Client: client}
+}
+
+// Build is a helper method to perform a Docker build against the
+// provided Docker client. It will load the image if not specified,
+// create a container if one does not already exist, and start a
+// container if the Dockerfile contains RUN commands. It will cleanup
+// any containers it creates directly, and set the e.Image.ID field
+// to the generated image.
+func (e *ClientExecutor) Build(r io.Reader, args map[string]string) error {
+	b := NewBuilder(e)
+	b.Args = args
+
+	if e.Excludes == nil {
+		excludes, err := ParseDockerignore(e.Directory)
+		if err != nil {
+			return err
+		}
+		e.Excludes = append(excludes, ".dockerignore", "Dockerfile")
+	}
+
+	// TODO: check the Docker daemon version (1.20 is required for Upload)
+
+	node, err := parser.Parse(r)
+	if err != nil {
+		return err
+	}
+
+	// identify the base image
+	from, err := b.From(node)
+	if err != nil {
+		return err
+	}
+	// load the image
+	if e.Image == nil {
+		if from == NoBaseImageSpecifier {
+			if runtime.GOOS == "windows" {
+				return fmt.Errorf("building from scratch images is not supported")
+			}
+			from, err = e.CreateScratchImage()
+			if err != nil {
+				return err
+			}
+			defer e.CleanupImage(from)
+		}
+		glog.V(4).Infof("Retrieving image %q", from)
+		e.Image, err = e.LoadImage(from)
+		if err != nil {
+			return err
+		}
+	}
+
+	// update the builder with any information from the image, including ONBUILD
+	// statements
+	if err := b.FromImage(e.Image, node); err != nil {
+		return err
+	}
+
+	b.RunConfig.Image = from
+	glog.V(4).Infof("step: FROM %s", from)
+
+	// create a container to execute in, if necessary
+	mustStart := b.RequiresStart(node)
+	if e.Container == nil {
+		opts := docker.CreateContainerOptions{
+			Config: &docker.Config{
+				Image: from,
+			},
+		}
+		if mustStart {
+			// TODO: windows support
+			opts.Config.Cmd = []string{"sleep 86400"}
```

Comment:
I'll make this configurable by the caller - the eventual solution is to
inject a stub binary

On Apr 27, 2016, at 4:42 AM, Michal Fojtik <notifications@github.com> wrote:

In pkg/util/docker/dockerfile/builder/client.go
<https://github.com/openshift/origin/pull/7823#discussion_r61222700>:

> +	}
> +
> +	b.RunConfig.Image = from
> +	glog.V(4).Infof("step: FROM %s", from)
> +
> +	// create a container to execute in, if necessary
> +	mustStart := b.RequiresStart(node)
> +	if e.Container == nil {
> +		opts := docker.CreateContainerOptions{
> +			Config: &docker.Config{
> +				Image: from,
> +			},
> +		}
> +		if mustStart {
> +			// TODO: windows support
> +			opts.Config.Cmd = []string{"sleep 86400"}

@smarterclayton <https://github.com/smarterclayton> how about sleep infinity
?

—
You are receiving this because you were mentioned.
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/7823/files/c107acb4f0baf738c416c767091e3e7ea14f8558#r61222700>